### PR TITLE
Version upgrade support for TKGs and TKGm clusters

### LINKED ISSUE #31: TMC terraform provider to support provisioned cluster's version upgrade
### Is your feature request related to a problem? Please describe.

TMC provisioned workload clusters support upgrading of the k8s version. This is possible by updating the version attribute under spec.distribution.
Presently the TMC provider is not able to support this state change of the cluster

### Describe the solution you'd like

Updating the distribution version of the provisioned cluster should trigger the cluster upgrade process

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by shobha2626**
1. **What this PR does / why we need it**:
    Version upgrade support for TKGs and TKGm clusters
2. **Which issue(s) this PR fixes**

        (optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged):

        Fixes #31 

3. **Additional information**

4. **Special notes for your reviewer**

### CODE REVIEW

**Comment by shreyassreenivas** [comment]
Mention the GitHub issue number which is fixing this

File: internal/resources/cluster/resource_cluster.go [hunk 1]
```diff
@@ -295,42 +295,63 @@ func resourceClusterDelete(_ context.Context, d *schema.ResourceData, m interfac
...
+		if newClusterGroupName.(string) != "" {
+			cluster.Spec.ClusterGroupName = newClusterGroupName.(string)
+		}
 	}
+}
 
+func withMetaUpdate(config authctx.TanzuContext, d *schema.ResourceData, cluster *clustermodel.VmwareTanzuManageV1alpha1ClusterCluster) {
 	if common.HasMetaChanged(d) {
-		meta := common.ConstructMeta(d)
+		getResp, err := config.TMCConnection.ClusterResourceService.ManageV1alpha1ClusterResourceServiceGet(constructFullname(d))
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:323]: why is this needed in withMetaUpdate

File: internal/resources/cluster/resource_cluster.go [hunk 2]
```diff
@@ -295,42 +295,63 @@ func resourceClusterDelete(_ context.Context, d *schema.ResourceData, m interfac
...
+	config := m.(authctx.TanzuContext)
+
+	// Get call to initialise the cluster struct
+	getResp, err := config.TMCConnection.ClusterResourceService.ManageV1alpha1ClusterResourceServiceGet(constructFullname(d))
+	if err != nil {
+		return diag.FromErr(errors.Wrapf(err, "Unable to get Tanzu Mission Control cluster entry, name : %s", d.Get(clusterNameKey)))
 	}
 
-	incomingCGName := d.Get(helper.GetFirstElementOf(SpecKey, clusterGroupKey))
+	var updates []func(authctx.TanzuContext, *schema.ResourceData, *clustermodel.VmwareTanzuManageV1alpha1ClusterCluster)
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:349]: define this function definition and give it a name

File: internal/resources/cluster/resource_cluster.go [hunk 3]
```diff
@@ -295,42 +295,63 @@ func resourceClusterDelete(_ context.Context, d *schema.ResourceData, m interfac
...
+
+	updates = append(updates, withMetaUpdate, withClusterGroupUpdate, withTKGsVsphereVersionUpdate, withTKGmVsphereVersionUpdate)
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:351]: instead of appending, can't we add all these entries during updates array initialisation

File: internal/resources/cluster/resource_cluster.go [hunk 4]
```diff
@@ -295,42 +295,63 @@ func resourceClusterDelete(_ context.Context, d *schema.ResourceData, m interfac
...
 
-	if incomingCGName.(string) != "" {
-		getResp.Cluster.Spec.ClusterGroupName = incomingCGName.(string)
+	for _, update := range updates {
+		update(config, d, getResp.Cluster)
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:354]: enhance the update function to return a boolean if there an update, 
and if there the boolean is set and then invoke the update function

**Comment by shreyassreenivas** [comment]
Please add log statements either debug or info which will give more information

File: internal/resources/cluster/resource_cluster.go (hunk 1 above)
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go]: Avoiding it 👍

File: internal/resources/cluster/resource_cluster.go (hunk 2 above)
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go]: Done

File: internal/resources/cluster/resource_cluster.go (hunk 3 above)
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go]: Done

File: internal/resources/cluster/resource_cluster.go (hunk 4 above)
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go]: Return bool by every function seems redundant, making use of a var instead

File: internal/resources/cluster/resource_cluster.go [hunk 5]
```diff
@@ -295,68 +300,99 @@ func resourceClusterDelete(_ context.Context, d *schema.ResourceData, m interfac
...
-	if err != nil {
-		return diag.FromErr(errors.Wrapf(err, "Unable to get Tanzu Mission Control cluster entry, name : %s", d.Get(clusterNameKey)))
+func withClusterGroupUpdate(d *schema.ResourceData, cluster *clustermodel.VmwareTanzuManageV1alpha1ClusterCluster) {
+	if d.HasChange(helper.GetFirstElementOf(SpecKey, clusterGroupKey)) {
+		newClusterGroupName := d.Get(helper.GetFirstElementOf(SpecKey, clusterGroupKey))
+		if newClusterGroupName.(string) != "" {
+			cluster.Spec.ClusterGroupName = newClusterGroupName.(string)
+		}
+
+		updateAvailable = true
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:330]: updateAvailable and the log statement should be inside the if condition... Even better flip the if condition and return

**Review by shreyassreenivas — CHANGES_REQUESTED** [review]

File: internal/resources/cluster/resource_cluster.go (hunk 5 above)
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go]: Done

File: internal/resources/cluster/resource_cluster.go [hunk 6]
```diff
@@ -295,68 +300,99 @@ func resourceClusterDelete(_ context.Context, d *schema.ResourceData, m interfac
...
-	case d.HasChange(helper.GetFirstElementOf(SpecKey, clusterGroupKey)):
-		updateRequired = true
+		log.Printf("[INFO] updating TKGm vSphere workload cluster version to %s", newVersion.(string))
 	}
+}
 
-	if !updateRequired {
-		return diags
+func withTKGsVsphereVersionUpdate(d *schema.ResourceData, cluster *clustermodel.VmwareTanzuManageV1alpha1ClusterCluster) {
+	if d.HasChange(helper.GetFirstElementOf(SpecKey, tkgServiceVsphereKey, distributionKey, versionKey)) {
```
[the remark this answers is not in the document]
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go:316]: Done

File: internal/resources/cluster/resource_cluster.go [hunk 7]
```diff
@@ -295,68 +300,99 @@ func resourceClusterDelete(_ context.Context, d *schema.ResourceData, m interfac
...
-	)
-	if err != nil {
-		return diag.FromErr(errors.Wrapf(err, "Unable to update Tanzu Mission Control cluster entry, name : %s", d.Get(clusterNameKey)))
+	updates := updateCheck{withMetaUpdate, withClusterGroupUpdate, withTKGsVsphereVersionUpdate, withTKGmVsphereVersionUpdate}
+
+	for _, update := range updates {
+		update(d, getResp.Cluster)
+	}
+
+	if updateAvailable {
```
[the remark this answers is not in the document]
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go:371]: It is set in the beginning of the function

File: internal/resources/cluster/resource_cluster.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:316]: I do not see the change

File: internal/resources/cluster/resource_cluster.go [hunk 8]
```diff
@@ -295,68 +300,102 @@ func resourceClusterDelete(_ context.Context, d *schema.ResourceData, m interfac
...
+			&clustermodel.VmwareTanzuManageV1alpha1ClusterRequest{
+				Cluster: getResp.Cluster,
+			},
+		)
+		if err != nil {
+			return diag.FromErr(errors.Wrapf(err, "Unable to update Tanzu Mission Control cluster entry, name : %s", d.Get(clusterNameKey)))
+		}
 	}
 
+	log.Printf("update successful")
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:382]: update the log statement

File: internal/resources/cluster/resource_cluster.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:371]: where are u setting it to true >

File: internal/resources/cluster/resource_cluster.go [hunk 9]
```diff
@@ -295,68 +300,102 @@ func resourceClusterDelete(_ context.Context, d *schema.ResourceData, m interfac
...
-		&clustermodel.VmwareTanzuManageV1alpha1ClusterRequest{
-			Cluster: getResp.Cluster,
-		},
-	)
-	if err != nil {
-		return diag.FromErr(errors.Wrapf(err, "Unable to update Tanzu Mission Control cluster entry, name : %s", d.Get(clusterNameKey)))
+	updates := updateCheck{withMetaUpdate, withClusterGroupUpdate, withTKGsVsphereVersionUpdate, withTKGmVsphereVersionUpdate}
+
+	for _, update := range updates {
+		update(d, getResp.Cluster)
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:368]: updateAvailable, should be set here

File: internal/resources/cluster/resource_cluster.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go:316]: Not flipping the if condition, instead - https://github.com/vmware/terraform-provider-tanzu-mission-control/pull/32#discussion_r847620662

File: internal/resources/cluster/resource_cluster.go (hunk 8 above)
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go:382]: Any better log message?

File: internal/resources/cluster/resource_cluster.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:371]: the above change is not reflected

File: internal/resources/cluster/resource_cluster.go (hunk 9 above)
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go:368]: It should be the same

File: internal/resources/cluster/resource_cluster.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go:371]: I see the changes in the diff of the PR

File: internal/resources/cluster/resource_cluster.go [hunk 10]
```diff
@@ -26,7 +26,12 @@ import (
 	"github.com/vmware-tanzu/terraform-provider-tanzu-mission-control/internal/resources/common"
 )
 
-type contextMethodKey struct{}
+type (
+	contextMethodKey struct{}
+	updateCheck      []func(*schema.ResourceData, *clustermodel.VmwareTanzuManageV1alpha1ClusterCluster)
+)
+
+var updateAvailable bool
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:34]: avoid using global variable

File: internal/resources/cluster/resource_cluster.go (hunk 8 above)
Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:382]: log.Printf("[INFO] update successful")

File: internal/resources/cluster/resource_cluster.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go:316]:
```suggestion
	if ! d.HasChange(helper.GetFirstElementOf(SpecKey, tkgServiceVsphereKey, distributionKey, versionKey)) {
	return 
	}
	// then add the update logic below
```

File: internal/resources/cluster/resource_cluster.go (hunk 8 above)
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go]: Done

File: internal/resources/cluster/resource_cluster.go (hunk 10 above)
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go]: Done

File: internal/resources/cluster/resource_cluster.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go:317]: It is leading to more than two return true/false statements, hence sticking to the same

File: internal/resources/cluster/resource_cluster.go [hunk 11]
```diff
@@ -295,68 +298,117 @@ func resourceClusterDelete(_ context.Context, d *schema.ResourceData, m interfac
...
+		_, err = config.TMCConnection.ClusterResourceService.ManageV1alpha1ClusterResourceServiceUpdate(
+			&clustermodel.VmwareTanzuManageV1alpha1ClusterRequest{
+				Cluster: getResp.Cluster,
+			},
+		)
+		if err != nil {
+			return diag.FromErr(errors.Wrapf(err, "Unable to update Tanzu Mission Control cluster entry, name : %s", d.Get(clusterNameKey)))
+		}
+
+		log.Printf("[INFO] update successful")
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:394]: nit : cluster update successful

Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go]: Done

File: internal/resources/cluster/resource_cluster.go [hunk 12]
```diff
@@ -295,68 +298,117 @@ func resourceClusterDelete(_ context.Context, d *schema.ResourceData, m interfac
...
 
-	_, err = config.TMCConnection.ClusterResourceService.ManageV1alpha1ClusterResourceServiceUpdate(
-		&clustermodel.VmwareTanzuManageV1alpha1ClusterRequest{
-			Cluster: getResp.Cluster,
-		},
-	)
+	cluster.Meta.Labels = objectMeta.Labels
+	cluster.Meta.Description = objectMeta.Description
+
+	log.Printf("[INFO] updating meta data")
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:360]: updating cluster metadata. 

You have used cluster in all the logs statement so please keep it consistent

**Review by Vasundharashukla — APPROVED** [review]
LGTM

File: internal/resources/cluster/resource_cluster.go (hunk 12 above)
Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go:360]: Will take up in the next PR
